### PR TITLE
[WIP] #1373 lazy fields

### PR DESCRIPTION
**Pull request by ya-goodfella**
A proof-of-concept implementation of #1373. Under construction, published as draft, beware the dragons.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,4 +1,3 @@
-import sbtcrossproject.CrossPlugin.autoImport.{CrossType, crossProject}
```
Comment (ya-goodfella) [inline build.sbt:1]: unused import

File: scalapb-runtime/src/main/scala/scalapb/LazyField.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package scalapb
+
+import com.google.protobuf.{ByteString, CodedOutputStream}
+
+import scala.language.implicitConversions
+
+sealed trait LazyField[+T] {
```
Comment (ya-goodfella) [inline scalapb-runtime/src/main/scala/scalapb/LazyField.scala:7]: The most intresting part goes there

File: scalapb-runtime/src/main/scala/scalapb/ByteStringDecoder.scala [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package scalapb
+
+import com.google.protobuf.ByteString
+
+@FunctionalInterface
+trait ByteStringDecoder[T] {
+  def apply(bs: ByteString): T
+}
+
+case object ByteStringDecoder {
```
Comment (thesamet) [inline scalapb-runtime/src/main/scala/scalapb/ByteStringDecoder.scala:10]: What's the intent for using `case object` here and not `object`? The code base uses `object` so for consistency I would prefer that here as well as in other location, unless of course there's a technical reason.

File: protobuf/scalapb/scalapb.proto [hunk 4]
```diff
@@ -284,6 +290,9 @@ message FieldOptions {
   // is not provided. This enables to emulate required fields in proto3.
   optional bool required = 31;
 
+  // If true, complex field would be lazy
+  optional bool lazy_field = 100;
```
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:294]: just `lazy` (we know it's a field :) )

File: protobuf/scalapb/scalapb.proto [hunk 5]
```diff
@@ -284,6 +290,9 @@ message FieldOptions {
   // is not provided. This enables to emulate required fields in proto3.
   optional bool required = 31;
 
+  // If true, complex field would be lazy
```
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:293]: What's the definition of complex? Maybe replace the word complex with its definition?

File: protobuf/scalapb/scalapb.proto [hunk 6]
```diff
@@ -177,6 +177,9 @@ message ScalaPbOptions {
   // If true, getters will be generated.
   optional bool getters = 27 [default=true];
 
+  // If true, complex fields would be lazy
+  optional bool lazy_fields = 100;
```
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:181]: can we remove this option and leave only the field-level option for now? Through field transformations feature users can specify rules that turn on `lazy` with more precision (only for repeated fields, only for certain types, etc).

File: protobuf/scalapb/scalapb.proto [hunk 7]
```diff
@@ -222,6 +225,9 @@ message MessageOptions {
   // file-level value, and can be overridden by the field-level setting.
   optional bool no_default_values_in_constructor = 9;
 
+  // If true, complex fields would be lazy
```
Comment (thesamet) [inline protobuf/scalapb/scalapb.proto:228]: see comment above on `lazy_field`

File: project/plugins.sbt [hunk 8]
```diff
@@ -36,8 +36,5 @@ addSbtPlugin("org.scalameta" % "sbt-mdoc" % "2.3.2")
 
 addSbtPlugin("org.scalameta" % "sbt-native-image" % "0.3.2")
 
-// For https://github.com/lightbend/mima/issues/422
-resolvers += Resolver.url(
-  "typesafe sbt-plugins",
-  url("https://dl.bintray.com/typesafe/sbt-plugins")
-)(Resolver.ivyStylePatterns)
+addSbtPlugin("org.jetbrains" % "sbt-ide-settings" % "1.0.0")
```
Comment (thesamet) [inline project/plugins.sbt:39]: The lines removed are fine; can the added jetbrains plugin be removed though?

File: project/project/metals.sbt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+// DO NOT EDIT! This file is auto-generated.
```
Comment (thesamet) [inline project/project/metals.sbt:1]: to be removed. isn't metals.sbt in our gitignore?

File: project/project/project/metals.sbt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+// DO NOT EDIT! This file is auto-generated.
```
Comment (thesamet) [inline project/project/project/metals.sbt:1]: to be removed. isn't metals.sbt in our gitignore?

File: protobuf/scalapb/scalapb.proto (hunk 4 above)
Comment (thesamet) [reply protobuf/scalapb/scalapb.proto:294]: can you use the smallest available tag number?

File: scalapb-runtime/src/main/scala/scalapb/CodedOutputStreamEncoder.scala [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  implicit def messageCodedOutputStreamEncoder[M <: GeneratedMessage]: CodedOutputStreamEncoder[M] = new CodedOutputStreamEncoder[M] with Predef {
+    override def serializedSizeNoTag(value: M): Int = value.serializedSize
+
+    override def writeTo(output: CodedOutputStream, value: M): Unit = value.writeTo(output)
+  }
+}
+
+trait CodedOutputStreamDecoder[T] {
```
Comment (thesamet) [inline scalapb-runtime/src/main/scala/scalapb/CodedOutputStreamEncoder.scala:31]: File name suggested only the Encoder would be here. Maybe split to two files or rename the file to something like `CodedOutputStreamFormatters`...

File: scalapb-runtime/src/main/scala/scalapb/CodedOutputStreamEncoder.scala [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        s"""ScalaPB generated decoder <$decoder> failed to parse data from the CodedInputStream.
+           |It may be a result of stream truncation or defect in the generated decoder.
+           |If you feel that this is a defect, please, fill the issue to the ScalaPB GitHub page:
+           |https://github.com/scalapb/ScalaPB/issues""".stripMargin
+      )
+      with DecodeException {
+    initCause(cause)
+  }
+
+  sealed trait Predef { self: Singleton with Serializable => }
```
Comment (thesamet) [inline scalapb-runtime/src/main/scala/scalapb/CodedOutputStreamEncoder.scala:74]: It looksl ike we're trying to abstract over failure handling here? Can you add comments on this?

File: compiler-plugin/src/main/scala/scalapb/compiler/FieldGenContext.scala [hunk 13]
```diff
@@ -0,0 +1,3 @@
+package scalapb.compiler
```
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/FieldGenContext.scala:1]: unused?

File: compiler-plugin/src/main/scala/scalapb/compiler/GeneratorPrinter.scala [hunk 14]
```diff
@@ -0,0 +1,3 @@
+package scalapb.compiler
```
Comment (thesamet) [inline compiler-plugin/src/main/scala/scalapb/compiler/GeneratorPrinter.scala:1]: unused?

**Comment by thesamet** [comment]
I've added some initial comments. I haven't read the whole thing, but if this works as a proof-of-concept then perhaps the added complexity isn't that bad!

**Comment by thesamet** [comment]
Closing the PR as the effort seem abandoned now. Feel free to kick off a new discussion around this if you're interested to continue working on this in the future.

**Pull request closed without merging**